migration (#168)
holepunchto/pear-docs: migration.md

* `Pear.reload` -> `location.reload`

### Compat Mode

While v1 Pear APIs will continue to function (with deprecation messages), for projects that need a legacy-bridge during overlapping transitional periods, compat-mode can be opted into which will

* Silence deprecation warnings
* Ensure legacy methods to continue to function after their removal for a longer time period
* Limit ability to use new APIs in future
* Eventually itself be deprecated & removed

Compat mode is a temporary transitioning utility, if usage can be avoided in favour of moving to the equivalent methods this avoids potential for legacy lock-in.

Enable with `<script>Pear.constructor.COMPAT = true</script>` at the top of the `<head>` element in the HTML entrypoint and `Pear.constructor.COMPAT = true` at the top of worker or terminal app entrypoints.

Compat mode isn't needed for premigration, only if time is needed for the application to transition to equivalent v2 APIs.

## Config Migration

The `package.json` `pear` field supplies configuration information for Pear applications.

* `pear.userAgent` -> `pear.gui.userAgent`
* `pear.gui.hideable` -> `pear.gui.closeHides`

### Route-Handling Migration

> If applications are performing route handling, in v2 it is necessary for an application opt-in to route handling.

Author: ```suggestion
If applications are performing route handling, in v2 it is necessary for an application to opt-in to route handling.
```

> In v1 when a `pear://` link is clicked outside of the application (eg in another app), or supplied to `pear run` the application is either opened, or sent a wakeup notification if already open.

Author: ```suggestion
In v1 when a `pear://` link is clicked outside of the application (e.g. in another app), or supplied to `pear run` the application is either opened, or sent a wakeup notification if already open.
```

When opened, `Pear.config.linkData` contains the pear:// link pathname with the leading `/` omitted. For example pear://runtime/pathname/to/some/where `linkData` would be `pathname/to/some/where`.

When woken, `Pear.updates((wakeup) => {})` provides `wakeup.linkData`.

> In v2 `linkData` is still available in both cases, but soft-deprecated in favour of v2 routing:

Author: ```suggestion
In v2 `linkData` is still available in both cases, but soft-deprecated in favor of v2 routing:
```

* `Pear.app.route` & `wakeup.route` - full pathname: pear://runtime/pathname/to/some/where -> `/pathname/to/some/where`.
* `Pear.app.fragment` & `wakeup.fragment` - location hash without `#`: pear://runtime/pathname#something -> `something`
* `Pear.app.query` & `wakeup.query` - query string without `?`. pear://runtime/pathname?some-query -> `some-query`

In v1 any link regardless of pathname ends up opening/waking the application. In v2, by default, pathnames are treated as sub-apps. This means `pear run` and `pear-run` can execute any valid file staged to a Pear application. Therefore in v2, it is necessary to opt-in via route mapping using `package.json` `pear.routes`. For example:

```json
{
  "pear": {
    "routes": {
      "/from/pathname/": "foo.js"
    }
  }
}
```

This configuration would route `pear://yourapp/from/pathname` to `pear://yourapp/foo.js` but its `Pear.app.route` / `wakeup.route` would be `/from/pathname`.

To opt-in fully to v1 behaviour the catch-all can be used:

```json
{
  "pear": {
    "routes": "."
  }
}
```

This configuration routes all pathnames to the application entrypoint.

> The `pear.unrouted` field may be used provide exceptions to the catch-all, for instance in order to run a sub-app (a "worker").

Author: ```suggestion
The `pear.unrouted` field may be used provide exceptions to the catch-all, for instance in order to run a sub-app (a worker).
```

# Migration

## Pear v1.x.x to Pear v2.x.x

* v1 supports `.html` entrypoints, v2 does not
* to migrate, use a `.js` entrypoint that uses [pear-electron](https://github.com/holepunchto/pear-electron) and [pear-bridge](https://github.com/holepunchto/pear-bridge)

### Modify `package.json`

* `package.json` `main` must point to a `.js` file instead of `.html`, e.g. `index.html` -> `index.js`
* `pear-electron` and `pear-bridge` must be specified in `dependencies`
* add a `bootstrap` field to `package.json` `scripts`: `"bootstrap": "pear-electron"`, optionally also do `"postinstall": "npm run bootstrap"`.

### Create JS entrypoint

```

> /** @typedef {import('pear-interface')} */

Reviewer: Is this line still needed? `Pear global` is no longer used in index.js it seems.
